### PR TITLE
CT-967: Fix for progress bar going over 100%

### PR DESCRIPTION
**Pull request by marticos**
https://limesurvey.atlassian.net/browse/MW-1354

### CODE REVIEW

File: application/models/services/SurveyDeactivate.php [hunk 1]
```diff
@@ -144,7 +144,10 @@ protected function archiveToken($iSurveyID, $date, $userID, $DBDate, &$aData)
             'token',
             $DBDate,
             $aData['aSurveysettings']['tokenencryptionoptions'],
-            json_encode($aData['aSurveysettings']['attributedescriptions'])
+            json_decode(
```
[the remark this answers is not in the document]
Comment (marticos) [reply application/models/services/SurveyDeactivate.php:147]: This json_decode function is meant to fix the "Invalid argument" error highlighted by the pipeline.
https://github.com/LimeSurvey/LimeSurvey/actions/runs/10925927891/job/30328601663

File: application/helpers/SurveyRuntimeHelper.php [hunk 2]
```diff
@@ -332,9 +332,15 @@ public function run($surveyid, $args)
                 $this->aSurveyInfo['progress']['currentstep'] = $_SESSION[$this->LEMsessid]['totalsteps'] + 1;
                 $this->aSurveyInfo['progress']['total']       = $totalVisibleSteps ? $totalVisibleSteps : $_SESSION[$this->LEMsessid]['totalsteps'];
             } else {
-                $this->aSurveyInfo['progress']['currentstep'] = $_SESSION[$this->LEMsessid]['step'];
+                $this->aSurveyInfo['progress']['currentstep'] = $_SESSION[$this->LEMsessid]['step'] - ($_SESSION[$this->LEMsessid]['hiddenSteps'] ?? 0);
```
[the remark this answers is not in the document]
Comment (marticos) [reply application/helpers/SurveyRuntimeHelper.php:335]: The hidden steps are define and calculated in application/helpers/expressions/em_manager_helper.php for example in the lines 4756 and 4789. Therefor they can be null or int.

File: application/models/services/SurveyDeactivate.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (lajosarpad) [reply application/models/services/SurveyDeactivate.php:147]: Like in the other PR, we need to make sure that the fix is correct by asking about it the author of that line.

**Comment by lajosarpad** [comment]
> > The issue did not reproduce for me on master. Did you test this? Also, https://github.com/LimeSurvey/LimeSurvey/pull/3957/files#diff-879ad87980b966696a7b02de0c55e0b02565637e26280383632f6c8081a9c30fR148 needs to be clarified.
> 
> Yes, I did test the fix. To reproduce the issue you need a survey with a few hidden questions and the display mode has to be question by question.
> 
> Here is a survey with hidden questions. [ [limesurvey_survey_274817 (2).lss.zip](https://github.com/user-attachments/files/17117341/limesurvey_survey_274817.2.lss.zip) ](url)
> 
> Thanks.

The survey I tested with was this one, but it did not surpass 100% for some reason. Nevertheless, if the test passed for you, then we just need to find out whether your change regarding JSON encoding seems to be good for the author of that code too.

**Comment by marticos** [comment]
> > > The issue did not reproduce for me on master. Did you test this? Also, https://github.com/LimeSurvey/LimeSurvey/pull/3957/files#diff-879ad87980b966696a7b02de0c55e0b02565637e26280383632f6c8081a9c30fR148 needs to be clarified.
> > 
> > 
> > Yes, I did test the fix. To reproduce the issue you need a survey with a few hidden questions and the display mode has to be question by question.
> > Here is a survey with hidden questions. [ [limesurvey_survey_274817 (2).lss.zip](https://github.com/user-attachments/files/17117341/limesurvey_survey_274817.2.lss.zip) ](url)
> > Thanks.
> 
> The survey I tested with was this one, but it did not surpass 100% for some reason. Nevertheless, if the test passed for you, then we just need to find out whether your change regarding JSON encoding seems to be good for the author of that code too.

OK, I would leave the proper testing to @tiborpacalat if he is fine with it and I will provide to him the survey made by Tim with the hidden questions and also the one of the customer that found the issue. That one was going over 100% for sure.
Thank you.

**Review by lajosarpad — APPROVED** [review]

**Pull request merged**